pkg/applicationserver/applicationserver.go

```diff
@@ -588,6 +592,202 @@ func (as *ApplicationServer) handleJoinAccept(ctx context.Context, ids ttnpb.End
+		return errPayloadCryptoDisabled
+	}
+	newCurrentQueue, newPendingQueue, err := f(ctx, dev)
+	if err != nil {
+		return err
+	}
+	downlinks := append(newCurrentQueue[:0:0], newCurrentQueue...)
+	downlinks = append(downlinks, newPendingQueue...)
+	client := ttnpb.NewAsNsClient(link.conn)
+	req := &ttnpb.DownlinkQueueRequest{
+		EndDeviceIdentifiers: dev.EndDeviceIdentifiers,
+		Downlinks:            downlinks,
+	}
+	_, err = client.DownlinkQueueReplace(ctx, req, link.callOpts...)
+	return err
+}
+
+var errUnknownSession = errors.DefineNotFound("unknown_session", "unknown session")
+
+// recomputePendingQueue partitions the invalidated downlinks based on their session key ID. Downlinks which are part of the
+// current session are not modified, while the downlinks from the previous pending session are migrated to the new pending
+// session.
+// This method requires the given end device's pending session to be set. This method mutates the end device's pending session LastAFCntDown.
+// This method does not change the contents of the given invalid downlink queue.
+func (as *ApplicationServer) recomputePendingQueue(ctx context.Context, dev *ttnpb.EndDevice, link *link, previousPendingSession *ttnpb.Session, invalidatedDownlinks []*ttnpb.ApplicationDownlink) error {
+	return as.changeDownlinkQueue(ctx, dev, link, func(ctx context.Context, dev *ttnpb.EndDevice) ([]*ttnpb.ApplicationDownlink, []*ttnpb.ApplicationDownlink, error) {
+		var previousPendingQueue, currentQueue []*ttnpb.ApplicationDownlink
+		unmatched := invalidatedDownlinks
+		if previousPendingSession != nil {
+			previousPendingQueue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(previousPendingSession.SessionKeyID, unmatched...)
+		}
+		if dev.Session != nil {
+			currentQueue, unmatched = ttnpb.PartitionDownlinksBySessionKeyIDEquality(dev.Session.SessionKeyID, unmatched...)
+		}
+		logger := log.FromContext(ctx)
+		for _, item := range unmatched {
+			logger.WithFields(log.Fields(
+				"f_port", item.FPort,
+				"f_cnt", item.FCnt,
+				"session_key_id", item.SessionKeyID,
+			)).Warn("Drop downlink message; unknown session key ID")
+			registerDropDownlink(ctx, dev.EndDeviceIdentifiers, item, errUnknownSession)
+		}
+
+		var sourceQueue []*ttnpb.ApplicationDownlink
+		var sourceSession *ttnpb.Session
+		switch {
+		case previousPendingSession == nil && dev.Session == nil:
+			// We cannot decode any of the downlinks if both sessions are missing.
+			return nil, nil, errNoDeviceSession
+		case dev.Session != nil:
+			// In the presence of a current session we copy the items from the current session to the pending session.
+			sourceQueue = currentQueue
+			sourceSession = dev.Session
+			logger.Debug("Initializing pending downlink queue from the current session")
+		case previousPendingSession != nil:
+			// In the presence of a previous session and the absence of a current session we migrate the items from the
+			// previous pending session to the new pending session.
```

Comment:
Shouldn't we take the `session`'s queue and put that in the new pending queue?

Reply:
Given the structure of the case, we entered the branch of `previousPendingSession != nil`, the branch of `dev.Session != nil` must be false, and as such there is no current session (or queue). If there is a current session, we indeed copy them to the queue of the pending session.